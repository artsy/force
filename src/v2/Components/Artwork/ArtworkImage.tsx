import { Image as BaseImage, ImageProps } from "@artsy/palette"
import styled from "styled-components"
import { MagnifyImage } from "../MagnifyImage"

interface ArtworkImageProps extends ImageProps {
  shouldZoomOnHover?: boolean
}

export const ArtworkImage: React.FC<ArtworkImageProps> = ({
  shouldZoomOnHover,
  ...rest
}) => {
  if (shouldZoomOnHover) {
    return <MagnifyImage {...rest} />
  }

  return <Image {...rest} />
}

const Image = styled(BaseImage)`
  display: block;
  width: 100%;
  height: 100%;
`
