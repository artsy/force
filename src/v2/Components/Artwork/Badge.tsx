import { Flex, Text } from "@artsy/palette"
import { Badge_artwork } from "v2/__generated__/Badge_artwork.graphql"
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components"
import { themeGet } from "@styled-system/theme-get"
import { useArtworkGridContext } from "../ArtworkGrid/ArtworkGridContext"

interface BadgeProps {
  artwork: Badge_artwork
  width?: number // for smaller images, we have a tweaked layout
}

const MIN_IMAGE_SIZE = 150

const Badge: React.FC<BadgeProps> = ({ artwork, width }) => {
  const { isAuctionArtwork } = useArtworkGridContext()

  if (isAuctionArtwork) {
    return null
  }

  const { is_biddable, sale } = artwork
  const includeBidBadge = is_biddable || (sale && sale.is_preview)
  // E.g.(ENDS IN 59M)
  const saleTimingHint =
    sale && sale.display_timely_at ? ` (${sale.display_timely_at})` : ""

  const isStackedLayout = (() => {
    // During the SSR render pass we don't have access to window pixel data so
    // default to high density screen.
    const devicePixelRatio =
      typeof window === "undefined" ? 2 : window.devicePixelRatio

    if (width) {
      return width / devicePixelRatio < MIN_IMAGE_SIZE
    } else {
      return false
    }
  })()

  return (
    <Badges flexDirection={isStackedLayout ? "column" : "row"}>
      {includeBidBadge && (
        <Label ml={1} mb={1}>
          Bid{saleTimingHint}
        </Label>
      )}
    </Badges>
  )
}

export default createFragmentContainer(Badge, {
  artwork: graphql`
    fragment Badge_artwork on Artwork {
      is_biddable: isBiddable
      href
      sale {
        is_preview: isPreview
        display_timely_at: displayTimelyAt
      }
    }
  `,
})

const Label = styled(Text)`
  background-color: ${themeGet("colors.white100")};
  border-radius: 2px;
  box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.1);
  font-size: 8px;
  letter-spacing: 0.3px;
  line-height: 1;
  padding: 2px 5px;
  text-transform: uppercase;
`

const Badges = styled(Flex)`
  position: absolute;
  bottom: 0;
  left: 0;
  pointer-events: none;
`
