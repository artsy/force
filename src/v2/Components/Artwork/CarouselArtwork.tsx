import { Box, Image } from "@artsy/palette"
import { useSystemContext } from "v2/Artsy"
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import { RouterLink } from "v2/Artsy/Router/RouterLink"
import { CarouselArtwork_artwork } from "v2/__generated__/CarouselArtwork_artwork.graphql"
import { ClickContainer, SaveButtonFragmentContainer } from "./SaveButton"
import Metadata from "./Metadata"
import { AuthContextModule } from "@artsy/cohesion"
import styled from "styled-components"
import { Flex } from "@artsy/palette"

interface CarouselArtworkProps {
  artwork: CarouselArtwork_artwork
  contextModule: AuthContextModule
  hideArtistName?: boolean
  hidePartnerName?: boolean
  hideSaleInfo?: boolean
  lazyLoad?: boolean
  showExtended?: boolean
  showMetadata?: boolean
  onClick?: () => void
}

const CarouselArtwork: React.FC<CarouselArtworkProps> = ({
  artwork,
  contextModule,
  hideArtistName,
  hidePartnerName,
  hideSaleInfo,
  lazyLoad,
  onClick,
  showExtended,
  showMetadata = true,
}) => {
  const { mediator, user } = useSystemContext()

  return (
    <Box>
      <RouterLink
        to={artwork.href}
        noUnderline
        onClick={() => {
          if (onClick) {
            onClick()
          }
        }}
      >
        <Container>
          <Image
            src={artwork.image.resized.src}
            srcSet={artwork.image.resized.srcSet}
            lazyLoad={lazyLoad}
          />

          <SaveButtonFragmentContainer
            mediator={mediator}
            user={user}
            contextModule={contextModule}
            artwork={artwork}
          />
        </Container>

        {showMetadata && (
          <Metadata
            artwork={artwork}
            extended={showExtended}
            hidePartnerName={hidePartnerName}
            hideArtistName={hideArtistName}
            hideSaleInfo={hideSaleInfo}
          />
        )}
      </RouterLink>
    </Box>
  )
}

const Container = styled(Flex)`
  position: relative;

  &:hover {
    ${ClickContainer} {
      opacity: 1;
    }
  }
`

export const CarouselArtworkFragmentContainer = createFragmentContainer(
  CarouselArtwork,
  {
    artwork: graphql`
      fragment CarouselArtwork_artwork on Artwork
        @argumentDefinitions(width: { type: "Int", defaultValue: 200 }) {
        image {
          resized(width: $width) {
            src
            srcSet
          }
          aspectRatio
          height
        }
        imageTitle
        title
        href
        ...Metadata_artwork
        ...SaveButton_artwork
        ...Badge_artwork
      }
    `,
  }
)
