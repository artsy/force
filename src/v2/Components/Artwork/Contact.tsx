import { Contact_artwork } from "v2/__generated__/Contact_artwork.graphql"
import * as React from "react";
import { createFragmentContainer, graphql } from "react-relay"
import { get } from "v2/Utils/get"
import TextLink from "../TextLink"

export interface ContactProps extends React.HTMLProps<Contact> {
  artwork: Contact_artwork
}

export class Contact extends React.Component<ContactProps, null> {
  contactLine() {
    const { artwork } = this.props
    if (artwork.sale && artwork.sale.is_auction) {
      return this.auctionLine()
    } else if (artwork.is_inquireable) {
      return this.contactPartnerLine()
    }
  }

  auctionLine() {
    const { artwork } = this.props
    // @ts-expect-error STRICT_NULL_CHECK
    const isLiveOpen = get(artwork, p => p.sale.is_live_open)
    // @ts-expect-error STRICT_NULL_CHECK
    const isOpen = get(artwork, p => p.sale.is_open)
    // @ts-expect-error STRICT_NULL_CHECK
    const isClosed = get(artwork, p => p.sale.is_closed)

    if (isLiveOpen) {
      return (
        // @ts-expect-error STRICT_NULL_CHECK
        <TextLink href={artwork.href} underline>
          Enter Live Auction
        </TextLink>
      )
    } else if (isOpen) {
      const sa = get(artwork, p => p.sale_artwork)
      // @ts-expect-error STRICT_NULL_CHECK
      const bidderPositions = get(sa, p => p.counts.bidder_positions)
      // @ts-expect-error STRICT_NULL_CHECK
      const highestBidDisplay = get(sa, p => p.highest_bid.display)
      // @ts-expect-error STRICT_NULL_CHECK
      const openingBidDisplay = get(sa, p => p.opening_bid.display)

      if (bidderPositions && bidderPositions > 0) {
        const s = bidderPositions > 1 ? "s" : ""
        return (
          <span>
            {highestBidDisplay} ({bidderPositions} bid
            {s})
          </span>
        )
      } else {
        return <span>{openingBidDisplay}</span>
      }
    } else if (isClosed) {
      return <span>Auction closed</span>
    } else {
      return <span />
    }
  }

  contactPartnerLine() {
    const partner = get(this.props, p =>
      // @ts-expect-error STRICT_NULL_CHECK
      p.artwork.partner.type.toLocaleLowerCase()
    )

    if (partner) {
      return (
        // @ts-expect-error STRICT_NULL_CHECK
        <TextLink href={this.props.artwork.href} underline>
          Contact {partner}
        </TextLink>
      )
    } else {
      return null
    }
  }

  render() {
    return <div>{this.contactLine()}</div>
  }
}

// @ts-expect-error STRICT_NULL_CHECK
export default createFragmentContainer(Contact, {
  artwork: graphql`
    fragment Contact_artwork on Artwork {
      href
      is_inquireable: isInquireable
      sale {
        is_auction: isAuction
        is_live_open: isLiveOpen
        is_open: isOpen
        is_closed: isClosed
      }
      partner(shallow: true) {
        type
      }
      sale_artwork: saleArtwork {
        highest_bid: highestBid {
          display
        }
        opening_bid: openingBid {
          display
        }
        counts {
          bidder_positions: bidderPositions
        }
      }
    }
  `,
})
