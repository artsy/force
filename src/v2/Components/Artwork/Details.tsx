import { Flex, Link, Text } from "@artsy/palette"
import { Details_artwork } from "v2/__generated__/Details_artwork.graphql"
import { SystemContextConsumer } from "v2/Artsy"
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components"
import { get } from "v2/Utils/get"

const TruncatedLine = styled.div`
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 1;
  -webkit-box-orient: vertical;
`

export interface Props extends React.HTMLProps<Details> {
  includeLinks: boolean
  artwork: Details_artwork
  hideSaleInfo?: boolean
  hideArtistName?: boolean
  hidePartnerName?: boolean
}

export class Details extends React.Component<Props, null> {
  static defaultProps = {
    includeLinks: true,
    hideSaleInfo: false,
  }

  artistLine() {
    const cultural_maker = this.props.artwork?.cultural_maker
    const artists = this.props.artwork?.artists
    const { includeLinks, hideArtistName } = this.props

    if (hideArtistName) {
      return
    }

    if (cultural_maker) {
      return (
        <TruncatedLine>
          <Text variant="mediumText">{cultural_maker}</Text>
        </TruncatedLine>
      )
    } else if (artists && artists.length) {
      const artistLine = artists
        .reduce((acc, artist, index) => {
          return acc.concat([
            ", ",
            includeLinks
              ? this.link(artist.name, artist.href, artist.id + "-" + index)
              : artist.name,
          ])
        }, [])
        .slice(1)
      return (
        <TruncatedLine>
          <Text variant="mediumText">{artistLine}</Text>
        </TruncatedLine>
      )
    }
  }

  titleLine() {
    const { includeLinks } = this.props
    const title = this.props?.artwork?.title
    const date = this.props?.artwork?.date
    const href = this.props?.artwork?.href

    const artworkText = (
      <TruncatedLine>
        <Text variant="text" color="black60">
          {date ? title + ", " + date : title}
        </Text>
      </TruncatedLine>
    )
    const artworkTextWithLink = <Link href={href}>{artworkText}</Link>

    return includeLinks ? artworkTextWithLink : artworkText
  }

  line(text) {
    return (
      <TruncatedLine>
        <Text variant="text" color="black60">
          {text}
        </Text>
      </TruncatedLine>
    )
  }

  link(text, href, key) {
    return (
      <Link href={href} key={key}>
        <Text variant="text" color="black60">
          {text}
        </Text>
      </Link>
    )
  }

  partnerLine() {
    const { hidePartnerName, includeLinks } = this.props
    const collecting_institution = this.props?.artwork?.collecting_institution
    const partner = this.props?.artwork?.partner
    const href = partner?.href
    const name = partner?.name

    if (hidePartnerName) {
      return
    }

    if (collecting_institution) {
      return this.line(collecting_institution)
    } else if (partner) {
      // TODO: We wrap the entire Metadata comp in an anchor tag linking to the artwork page, so why is there a link here?
      if (includeLinks) {
        return (
          <TruncatedLine>
            <Link href={href}>
              <Text variant="text" color="black60">
                {name}
              </Text>
            </Link>
          </TruncatedLine>
        )
      } else {
        return this.line(name)
      }
    }
  }

  saleInfoLine() {
    const { hideSaleInfo } = this.props
    if (hideSaleInfo) {
      return null
    }
    return (
      <Flex>
        <TruncatedLine>
          <Text variant="text" color="black60">
            {this.saleMessage()} {this.bidInfo()}
          </Text>
        </TruncatedLine>
      </Flex>
    )
  }

  bidInfo() {
    const { artwork } = this.props
    const { sale } = this.props.artwork

    const inRunningAuction = sale && sale.is_auction && !sale.is_closed
    if (!inRunningAuction) {
      return undefined
    }

    const bidderPositionCounts = get(
      artwork,
      a => a.sale_artwork.counts.bidder_positions,
      0
    )

    if (bidderPositionCounts === 0) {
      return undefined
    }

    const s = bidderPositionCounts > 1 ? "s" : ""
    return `(${bidderPositionCounts} bid${s})`
  }

  saleMessage() {
    const { artwork } = this.props
    const { sale } = artwork
    const isAuction = sale && sale.is_auction

    if (isAuction) {
      const showBiddingClosed = sale.is_closed
      if (showBiddingClosed) {
        return "Bidding closed"
      } else {
        const highestBidDisplay = get(
          artwork,
          p => p.sale_artwork.highest_bid.display
        )
        const openingBidDisplay = get(
          artwork,
          p => p.sale_artwork.opening_bid.display
        )

        return highestBidDisplay || openingBidDisplay || ""
      }
    }

    // TODO: Extract this sentence-cased version and apply everywhere.
    if (artwork.sale_message === "Contact For Price") {
      return "Contact for price"
    }

    return artwork.sale_message
  }

  render() {
    return (
      <SystemContextConsumer>
        {({ user }) => {
          return (
            <div>
              {this.artistLine()}
              {this.titleLine()}
              {this.partnerLine()}
              {this.saleInfoLine()}
            </div>
          )
        }}
      </SystemContextConsumer>
    )
  }
}

export const DetailsFragmentContainer = createFragmentContainer(Details, {
  artwork: graphql`
    fragment Details_artwork on Artwork {
      href
      title
      date
      sale_message: saleMessage
      cultural_maker: culturalMaker
      artists(shallow: true) {
        id
        href
        name
      }
      collecting_institution: collectingInstitution
      partner(shallow: true) {
        name
        href
      }
      sale {
        is_auction: isAuction
        is_closed: isClosed
      }
      sale_artwork: saleArtwork {
        counts {
          bidder_positions: bidderPositions
        }
        highest_bid: highestBid {
          display
        }
        opening_bid: openingBid {
          display
        }
      }
    }
  `,
})
