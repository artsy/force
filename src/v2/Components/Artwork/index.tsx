import { Artwork_artwork } from "v2/__generated__/Artwork_artwork.graphql"
import React from "react"
// @ts-ignore
import { ComponentRef, createFragmentContainer, graphql } from "react-relay"
import styled, { css } from "styled-components"
import theme from "../../Assets/Theme"
import Metadata from "./Metadata"

const Container = styled.div`
  width: 100%;
`

const ImageContainer = styled.div`
  width: 100%;
  position: relative;
  cursor: pointer;

  &::before {
    content: "";
    display: block;
    padding-bottom: 100%;
  }

  & .overlay-container {
    transition: opacity 0.25s;

    &.hovered {
      opacity: 0;
      visibility: hidden;
    }

    &.selected {
      opacity: 1;
      visibility: visible;
    }
  }

  &:hover {
    & .overlay-container.hovered {
      ${css`
        @media (min-width: ${theme.flexboxgrid.breakpoints.sm + 1}px) {
          opacity: 1;
          visibility: visible;
        }
      `};
    }
  }
`

const Image = styled.img`
  max-width: 100%;
  max-height: 100%;
  margin: auto;
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  right: 0;
`

export interface OverlayProps {
  selected: boolean
}

export interface ArtworkProps {
  extended?: boolean
  Overlay?: React.SFC<OverlayProps>
  artwork: Artwork_artwork
  onSelect?: (selected: boolean) => void
  showOverlayOnHover?: boolean
}

export interface ArtworkState {
  isSelected: boolean
}

export class Artwork extends React.Component<ArtworkProps, ArtworkState> {
  static defaultProps = {
    extended: true,
    overlay: null,
    showOverlayOnHover: false,
  }

  state = {
    isSelected: false,
  }

  onSelected = () => {
    if (!this.props.Overlay) {
      return
    }

    this.setState({
      isSelected: !this.state.isSelected,
    })

    if (this.props.onSelect) {
      this.props.onSelect(!this.state.isSelected)
    }
  }

  render() {
    const { artwork, Overlay, showOverlayOnHover } = this.props
    let overlayClasses = "overlay-container"

    overlayClasses += showOverlayOnHover ? " hovered" : ""
    overlayClasses += this.state.isSelected ? " selected" : ""

    return (
      <Container onClick={this.onSelected}>
        <ImageContainer>
          {/* @ts-expect-error STRICT_NULL_CHECK */}
          <Image src={artwork.image.url} />
          <div className={overlayClasses}>
            {Overlay && <Overlay selected={this.state.isSelected} />}
          </div>
        </ImageContainer>
        <Metadata extended={this.props.extended} artwork={artwork} />
      </Container>
    )
  }
}

export default createFragmentContainer(Artwork, {
  artwork: graphql`
    fragment Artwork_artwork on Artwork {
      slug
      image {
        url(version: "large")
        aspect_ratio: aspectRatio
      }
      ...Metadata_artwork
    }
  `,
})
