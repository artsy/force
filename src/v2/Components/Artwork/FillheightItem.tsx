import { AuthContextModule } from "@artsy/cohesion"
import { Box, Image } from "@artsy/palette"
import { FillheightItem_artwork } from "v2/__generated__/FillheightItem_artwork.graphql"
import { SystemContextProps } from "v2/Artsy"
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components"
import { getENV } from "v2/Utils/getENV"
import { userIsTeam } from "v2/Utils/user"
import Badge from "./Badge"
import Metadata from "./Metadata"
import {
  ClickContainer,
  SaveButtonFragmentContainer as SaveButton,
} from "./SaveButton"
import { RouterLink } from "v2/Artsy/Router/RouterLink"
import { useSystemContext } from "@artsy/reaction/dist/Artsy"
import { Mediator } from "lib/mediator"

const IMAGE_QUALITY = 80

const Placeholder = styled(Box).attrs({ bg: "gray10" })`
  position: relative;
  width: 100%;
`

interface FillheightItemContainerProps
  extends SystemContextProps,
    React.HTMLProps<typeof FillheightItemContainer> {
  artwork: FillheightItem_artwork
  contextModule: AuthContextModule
  hideArtistName?: boolean
  hidePartnerName?: boolean
  hideSaleInfo?: boolean
  imageWidth: number
  imageHeight: number
  lazyLoad?: boolean
  marginLeft?: number
  marginRight?: number
  showExtended?: boolean
  showMetadata?: boolean
  onClick?: () => void
}

const FillheightItemContainer: React.FC<FillheightItemContainerProps> = ({
  showMetadata = true,
  imageWidth = 220,
  artwork,
  className,
  contextModule,
  hideArtistName,
  hidePartnerName,
  hideSaleInfo,
  imageHeight,
  lazyLoad,
  onClick,
  showExtended,
  user,
}) => {
  const { mediator } = useSystemContext()

  if (!artwork.image) {
    return null
  }

  const getAspectRatio = () => {
    let aspectRatio = artwork?.image?.aspectRatio || 1
    // Our relay testing layer often mocks props with strings; this makes it a
    // little safer and less confusing when that happens.
    if (typeof aspectRatio !== "number") {
      aspectRatio = 1
    }
    return aspectRatio
  }

  const getImageHeight = () => {
    // NOTE: `aspectRatio` will default to `1` if image geometry is missing.
    // Given that `imageHeight` is required, this means that any missing geometry
    // will default to a square thumbnail.
    const height = Math.floor(imageWidth * getAspectRatio())
    return height
  }

  const getImageUrl = (pixelRatio = 1) => {
    if (!artwork.image.url) {
      return null
    }

    // Either scale or crop, based on if an aspect ratio is 1. Either the image
    // actually is a square â€” or the geometry is missing and utilizing fill prevents
    // distortion when it is sized to fit.
    const aspectRatio = getAspectRatio()
    const type = aspectRatio === 1 ? "fill" : "fit"
    const w = imageWidth * pixelRatio
    const h = getImageHeight() * pixelRatio

    const imageUrl = `${getENV(
      "GEMINI_CLOUDFRONT_URL"
    )}/?resize_to=${type}&width=${w}&height=${h}&quality=${IMAGE_QUALITY}&src=${encodeURIComponent(
      artwork.image.url
    )}`

    return imageUrl
  }

  const height = getImageHeight()

  return (
    <Box className={className} width={imageWidth}>
      <Placeholder style={{ height, width: imageWidth }}>
        <RouterLink
          to={artwork.href}
          onClick={() => {
            if (onClick) {
              onClick()
            }
          }}
        >
          <Image
            src={getImageUrl(1)}
            srcSet={`${getImageUrl(1)} 1x, ${getImageUrl(2)} 2x`}
            width={imageWidth}
            height={imageHeight}
            lazyLoad={lazyLoad}
            preventRightClick={!userIsTeam(user)}
            alt={artwork?.imageTitle}
          />
        </RouterLink>

        {showExtended && <Badge artwork={artwork} width={imageWidth} />}

        <SaveButton
          user={user}
          mediator={mediator as Mediator}
          contextModule={contextModule}
          artwork={artwork}
        />
      </Placeholder>

      {showMetadata && (
        <Metadata
          artwork={artwork}
          extended={showExtended}
          hidePartnerName={hidePartnerName}
          hideArtistName={hideArtistName}
          hideSaleInfo={hideSaleInfo}
        />
      )}
    </Box>
  )
}

const FillheightItem = styled(FillheightItemContainer)<
  FillheightItemContainerProps
>`
  display: inline-block;
  width: ${props => props.width}px;
  vertical-align: top;
  margin-right: ${props => props.marginRight || 0}px;
  margin-left: ${props => props.marginLeft || 0}px;

  &:hover {
    ${ClickContainer} {
      opacity: 1;
    }
  }
`

export const FillheightItemFragmentContainer = createFragmentContainer(
  FillheightItem,
  {
    artwork: graphql`
      fragment FillheightItem_artwork on Artwork {
        image {
          url(version: "large")
          aspectRatio
          height
        }
        imageTitle
        title
        href
        ...Metadata_artwork
        ...SaveButton_artwork
        ...Badge_artwork
      }
    `,
  }
)
