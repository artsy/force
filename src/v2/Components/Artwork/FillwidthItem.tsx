import { AuthContextModule } from "@artsy/cohesion"
import { Box, Image } from "@artsy/palette"
import { FillwidthItem_artwork } from "v2/__generated__/FillwidthItem_artwork.graphql"
import { SystemContextProps, withSystemContext } from "v2/Artsy"
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components"
import { get } from "v2/Utils/get"
import { getENV } from "v2/Utils/getENV"
import createLogger from "v2/Utils/logger"
import { userIsTeam } from "v2/Utils/user"
import Badge from "./Badge"
import Metadata from "./Metadata"
import {
  ClickContainer,
  SaveButtonFragmentContainer as SaveButton,
} from "./SaveButton"
import { RouterLink } from "v2/Artsy/Router/RouterLink"
import { Mediator } from "lib/mediator"

const logger = createLogger("FillwidthItem.tsx")

const IMAGE_QUALITY = 80

const Placeholder = styled(Box).attrs({ bg: "gray10" })`
  position: relative;
  width: 100%;
`

export interface FillwidthItemContainerProps
  extends SystemContextProps,
    React.HTMLProps<FillwidthItemContainer> {
  artwork: FillwidthItem_artwork
  contextModule: AuthContextModule
  hideArtistName?: boolean
  hidePartnerName?: boolean
  hideSaleInfo?: boolean
  imageHeight: number
  lazyLoad?: boolean
  marginLeft?: number
  marginRight?: number
  mediator?: Mediator
  showExtended?: boolean
  showMetadata?: boolean
  onClick?: () => void
}

export class FillwidthItemContainer extends React.Component<
  FillwidthItemContainerProps
> {
  static defaultProps = {
    showMetadata: true,
  }

  get imageWidth() {
    const {
      artwork: {
        image: { aspectRatio },
      },
    } = this.props

    // NOTE: `aspectRatio` will default to `1` if image geometry is missing.
    // Given that `imageHeight` is required, this means that any missing geometry
    // will default to a square thumbnail.
    return Math.floor(this.props.imageHeight * aspectRatio)
  }

  get imageHeight() {
    return this.props.imageHeight
  }

  getImageUrl(pixelRatio = 1) {
    const imageURL = this.props.artwork.image.url

    if (!imageURL) {
      return null
    }

    const {
      artwork: {
        image: { aspectRatio },
      },
    } = this.props

    // Either scale or crop, based on if an aspect ratio is 1. Either the image
    // actually is a square â€” or the geometry is missing and utilizing fill prevents
    // distortion when it is sized to fit.
    const type = aspectRatio === 1 ? "fill" : "fit"

    return `${getENV("GEMINI_CLOUDFRONT_URL")}/?resize_to=${type}&width=${
      this.imageWidth * pixelRatio
    }&height=${
      this.imageHeight * pixelRatio
    }&quality=${IMAGE_QUALITY}&src=${encodeURIComponent(imageURL)}`
  }

  render() {
    const {
      artwork,
      className,
      contextModule,
      imageHeight,
      lazyLoad,
      mediator,
      showExtended,
      showMetadata,
      user,
      hidePartnerName,
      hideArtistName,
      hideSaleInfo,
    } = this.props

    let userSpread = {}
    if (user) {
      userSpread = { user }
    }
    const isTeam = userIsTeam(user)

    const image = get(this.props, p => p.artwork.image)

    if (!image) {
      const href = get(this.props, p => p.artwork.href, "(unknown href)")
      logger.error(`Artwork at ${href} does not have an image!`)
      return null
    }

    return (
      <Box className={className} width={this.imageWidth}>
        <Placeholder
          style={{ height: this.imageHeight, width: this.imageWidth }}
        >
          <RouterLink
            to={artwork.href}
            onClick={() => {
              if (this.props.onClick) {
                this.props.onClick()
              }
            }}
          >
            <Image
              src={this.getImageUrl(1)}
              srcSet={`${this.getImageUrl(1)} 1x, ${this.getImageUrl(2)} 2x`}
              width="100%"
              height={imageHeight}
              lazyLoad={lazyLoad}
              preventRightClick={!isTeam}
              alt={artwork?.imageTitle}
            />
          </RouterLink>

          {showExtended && <Badge artwork={artwork} width={this.imageWidth} />}

          <SaveButton
            {...userSpread}
            mediator={mediator}
            contextModule={contextModule}
            artwork={artwork}
          />
        </Placeholder>

        {showMetadata && (
          <Metadata
            artwork={artwork}
            extended={showExtended}
            hidePartnerName={hidePartnerName}
            hideArtistName={hideArtistName}
            hideSaleInfo={hideSaleInfo}
          />
        )}
      </Box>
    )
  }
}

export const FillwidthItem = styled(FillwidthItemContainer)<
  FillwidthItemContainerProps
>`
  display: inline-block;
  width: ${props => props.width}px;
  vertical-align: top;
  margin-right: ${props => props.marginRight || 0}px;
  margin-left: ${props => props.marginLeft || 0}px;

  &:hover {
    ${ClickContainer} {
      opacity: 1;
    }
  }
`

export default createFragmentContainer(withSystemContext(FillwidthItem), {
  artwork: graphql`
    fragment FillwidthItem_artwork on Artwork {
      image {
        url(version: "large")
        aspectRatio
      }
      imageTitle
      title
      href
      ...Metadata_artwork
      ...SaveButton_artwork
      ...Badge_artwork
    }
  `,
})
