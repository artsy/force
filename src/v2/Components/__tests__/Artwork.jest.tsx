import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import Artwork from "../Artwork"

describe("Artwork", () => {
  it("renders correctly", () => {
    const artworkProps = {
      artists: [
        {
          id: "mikael-olson",
          name: "Mikael Olson",
        },
      ],
      date: "2015",
      href: "/artwork/mikael-olson-some-kind-of-dinosaur",
      id: "mikael-olson-some-kind-of-dinosaur",
      image: {
        aspect_ratio: 0.74,
        placeholder: 200,
        url: "artsy.net/image-url",
      },
      is_in_auction: false,
      partner: {
        name: "Gallery 1261",
      },
      sale_message: "$875",
      title: "Some Kind of Dinosaur",
    }

    const artwork = renderer
      .create(<Artwork artwork={artworkProps as any} />)
      .toJSON()
    expect(artwork).toMatchSnapshot()
  })
})
