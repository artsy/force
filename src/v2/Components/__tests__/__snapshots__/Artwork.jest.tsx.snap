// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Artwork renders correctly 1`] = `
.c8 {
  margin-bottom: 0.3px;
}

.c6 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
  color: black100;
}

.c7 {
  font-family: Unica77LLWebRegular,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 14px;
  line-height: 20px;
  color: #000;
}

.c3 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 15px;
  line-height: 1.25em;
  -webkit-font-smoothing: antialiased;
  color: #333333;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 {
  display: block;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c4 {
  font-family: Unica77LLWebRegular,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
  color: #666666;
  margin-top: 12px;
  text-align: left;
}

.c0 {
  width: 100%;
}

.c1 {
  width: 100%;
  position: relative;
  cursor: pointer;
}

.c1::before {
  content: "";
  display: block;
  padding-bottom: 100%;
}

.c1 .overlay-container {
  -webkit-transition: opacity 0.25s;
  transition: opacity 0.25s;
}

.c1 .overlay-container.hovered {
  opacity: 0;
  visibility: hidden;
}

.c1 .overlay-container.selected {
  opacity: 1;
  visibility: visible;
}

.c2 {
  max-width: 100%;
  max-height: 100%;
  margin: auto;
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

@media (min-width:769px) {
  .c1:hover .overlay-container.hovered {
    opacity: 1;
    visibility: visible;
  }
}

<div
  className="c0"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    <img
      className="c2"
      src="artsy.net/image-url"
    />
    <div
      className="overlay-container"
    />
  </div>
  <a
    className="c3"
    href="/artwork/mikael-olson-some-kind-of-dinosaur"
    target="_self"
  >
    <div
      className="c4"
    >
      <div>
        <div
          className="c5"
        >
          <div
            className="c6"
            color="black100"
            fontFamily={
              Object {
                "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
                "fontWeight": 500,
              }
            }
            fontSize="14px"
            style={Object {}}
          >
            Mikael Olson
          </div>
        </div>
        <div
          className="c5"
        >
          <span>
            Some Kind of Dinosaur
          </span>
          , 2015
        </div>
        <div
          className="c5"
        >
          Gallery 1261
        </div>
        <div
          className="c7"
          color="#000"
          fontFamily="Unica77LLWebRegular, 'Helvetica Neue', Helvetica, Arial, sans-serif"
          fontSize="14px"
          style={
            Object {
              "display": "inline",
            }
          }
        >
          $875
           
        </div>
        <div
          className="c7"
          color="#000"
          fontFamily="Unica77LLWebRegular, 'Helvetica Neue', Helvetica, Arial, sans-serif"
          fontSize="14px"
          style={
            Object {
              "display": "inline",
            }
          }
        />
        <div
          className="c8"
        />
      </div>
    </div>
  </a>
</div>
`;
